man/4s-import.1
.Dd 2009-07-08
.Dt 4S-IMPORT 1J 4store
.Os 4store
.Sh NAME
.Nm 4s-import
.Nd Import RDF into a 4store KB
.Sh SYNOPSIS
.Nm
kbname
.Op Fl v
.Op Fl a
.Op Fl M Ar default-model
.Op Fl m Ar model
.Op Fl f Ar format
.Op rdf-file
.Ar ...
.Bl -tag -width indent
.It Fl "v, \-\-verbose"
Increase verbosity (by default success is silent)
.It Fl "a, \-\-add"
Ordinarily importing RDF to a model replaces any previous contents of that model, by using the \-\-add flag this behaviour is overridden and any triples which were previously in the model are retained.
.It Fl "M, \-\-model-default"
Set a model (graph) URI which is used by default for all imported files
.It Fl "m, \-\-model"
Set a model (graph) URI for the next named file only (overrides \-M if it has been used)
.It Fl "f, \-\-format"
Tell the RDF parser the format of the files (if not specified the parser will guess)
.El
.Sh SEE ALSO
4s-query(1), 4s-size(1), 4s-httpd(1), 4s-backend(1), 4s-delete-model(1)
.Sh EXAMPLES
$
.Nm
demo one.ttl two.xrdf
.sp
Imports two files into the demo KB, each as a separate model (graph) guessing the RDF format used for each file
.sp
$
.Nm
demo
.Fl M
http://example.com/menu
starters.ttl mains.ttl sweets.ttl
.sp
Imports three files into a single graph named <http://example.com/menu>


NEWS
For up to date news see http://twitter.com/4store

Changes in 1.1.5

New features

* Non MDNS admin system, for clusters with lots of KBs, and systems with no
  multicast (e.g. Amazon EC2) — Dave Challis
* RDF 1.1 bNode Skolemsation (enabled with --enable-skolemization)
* Implementation of IN and NOT IN
* Increase maximum number of segments in one DB
* OFFSET and LIMIT now works with GROUP BY
* New functions: STRBEFORE, STRAFTER, UUID, STRUUID
* BIND implemented
* Chache stats in HTTP server status page
* Support skolemisation in CONSTRUCT and DESCRIBE
* Some support for xsd:date
* Access control support (graph level), see
  http://4store.org/trac/wiki/GraphAccessControl — Manuel Salvadores
* DISTINCT in Aggregates

Bugfixes

* Loads of improvements to test system — everyone
* Fix bug in disjunctive FILTER optimisation
* Handle Turtle MIME type correctly
* Bugfix for DESCRIBE, returning wrong RDF syntax — Mischa Tuffield
* Bugfix for DESCRIBE in graphs with cycles - Manuel Salvadores
* Bugfix to URI prefix compression - Manuel Salvadores
* Bugfix for numeric types in INSERT/DELETE DATA
* Better XML result output
* Bind all variables when WHERE pattern forms a triangle of triples - Manuel
  Salvadores
* Fix for shell tools — Pavol Rusnak
* Make FROM work with DESCRIBE
* Better portability across Linux distros — Rob Syme, Leif Warner
* SPARQL Update and language tag fixes
* Various optimiser improvments
* INSERT ... WHERE ... bugfixes
* ORDER BY on AS projected variables — Manuel Salvadores
* Better static build tools — Nicholas Humfrey


src/frontend/4s-import.c
/*
    4store - a clustered RDF storage and query engine

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
/*
 *  Copyright (C) 2006 Steve Harris for Garlik
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/types.h>
#include <sys/time.h>
#include <raptor.h>
#include <getopt.h>
#include <glib.h>
#include <fcntl.h>
#include <errno.h>

#include "../common/params.h"
#include "../common/4store.h"
#include "import.h"
#include "../common/error.h"
#include "../common/4store.h"

static fsp_link *fsplink; /* link to remote server(s) */

struct timeval then, then_last, now;

int main(int argc, char *argv[])
{
    int verbosity = 0;
    int dryrun = 0;
    char *password = NULL;
    char *format = "auto";
    FILE *msg = stderr;
    char *optstring = "ac:m:M:vnf:";
    int c, opt_index = 0, help = 0;
    int files = 0, adding = 0;
    char *kb_name = NULL;
    char *model[argc], *uri[argc];
    char *model_default = NULL;

    password = fsp_argv_password(&argc, argv);

    static struct option long_options[] = {
        { "add", 0, 0, 'a' },
        { "config-file", 1, 0, 'c' },
        { "model", 1, 0, 'm' },
        { "model-default", 1, 0, 'M' },
        { "verbose", 0, 0, 'v' },
        { "dryrun", 0, 0, 'n' },
        { "no-resources", 0, 0, 'R' },
        { "no-quads", 0, 0, 'Q' },
        { "format", 1, 0, 'f' },
        { "help", 0, 0, 'h' },
        { "version", 0, 0, 'V' },
        { 0, 0, 0, 0 }
    };

    for (int i= 0; i < argc; ++i) {
      model[i] = NULL;
    }

    int help_return = 1;

    while ((c = getopt_long (argc, argv, optstring, long_options, &opt_index)) != -1) {
        if (c == 'm') {
            model[files++] = optarg;
        } else if (c == 'M') {
            model_default = optarg;
        } else if (c == 'v') {
            verbosity++;
        } else if (c == 'a') {
            adding = 1;
        } else if (c == 'c') {
            fs_set_config_file(optarg); 
        } else if (c == 'n') {
            dryrun |= FS_DRYRUN_DELETE | FS_DRYRUN_RESOURCES | FS_DRYRUN_QUADS;
        } else if (c == 'R') {
            dryrun |= FS_DRYRUN_RESOURCES;
        } else if (c == 'Q') {
            dryrun |= FS_DRYRUN_QUADS;
        } else if (c == 'f') {
            format = optarg;
        } else if (c == 'h') {
            help = 1;
            help_return = 0;
        } else if (c == 'V') {
            printf("%s, built for 4store %s\n", argv[0], GIT_REV);
            exit(0);
        } else {
            help = 1;
        }
    }

    if (verbosity > 0) {
	if (dryrun & FS_DRYRUN_DELETE) {
	    printf("warning: not deleting old model\n");
	}
	if (dryrun & FS_DRYRUN_RESOURCES) {
	    printf("warning: not importing resource nodes\n");
	}
	if (dryrun & FS_DRYRUN_QUADS) {
	    printf("warning: not importing quad graph\n");
	}
    }

    files = 0;
    for (int k = optind; k < argc; ++k) {
        if (!kb_name) {
            kb_name = argv[k];
        } else {
	    if (strchr(argv[k], ':')) {
		uri[files] = g_strdup(argv[k]);
	    } else {
		uri[files] = (char *)raptor_uri_filename_to_uri_string(argv[k]);
	    }
            if (!model[files]) {
                if (!model_default) {
                    model[files] = uri[files];
                } else {
                    model[files] = model_default;
                }
            }
            files++;
        }
    }

    raptor_world *rw = raptor_new_world();
    if (help || !kb_name || files == 0) {
        fprintf(stdout, "%s revision %s\n", argv[0], FS_FRONTEND_VER);
        fprintf(stdout, "Usage: %s <kbname> <rdf file/URI> ...\n", argv[0]);
        fprintf(stdout, " -v --verbose   increase verbosity (can repeat)\n");
        fprintf(stdout, " -a --add       add data to models instead of replacing\n");
        fprintf(stdout, " -m --model     specify a model URI for the next RDF file\n");
        fprintf(stdout, " -M --model-default specify a model URI for all RDF files\n");
        fprintf(stdout, " -f --format    specify an RDF syntax for the import\n");
        fprintf(stdout, "\n   available formats are:\n");

        for (unsigned int i=0; 1; i++) {
            const raptor_syntax_description *desc =
                    raptor_world_get_parser_description(rw, i);
            if (!desc) {
                break;
            }
            fprintf(stdout, "    %12s - %s\n", desc->names[0], desc->label);
        }
        exit(help_return);
    }

    fsp_syslog_enable();

    fsplink = fsp_open_link(kb_name, password, FS_OPEN_HINT_RW);

    if (!fsplink) {
      fs_error (LOG_ERR, "couldn't connect to “%s”", kb_name);
      exit(2);
    }

    const char *features = fsp_link_features(fsplink);
    int has_o_index = !(strstr(features, "no-o-index")); /* tweak */

    fs_hash_init(fsp_hash_type(fsplink));
    const int segments = fsp_link_segments(fsplink);
    int total_triples = 0;

    fs_import_timing timing[segments];

    for (int seg = 0; seg < segments; seg++) {
        fsp_get_import_times(fsplink, seg, &timing[seg]);
    }

    gettimeofday(&then, 0);

    if (fsp_start_import_all(fsplink)) {
	fs_error(LOG_ERR, "aborting import");

	exit(3);
    }

    fs_rid_vector *mvec = fs_rid_vector_new(0);

    for (int f= 0; f < files; ++f) {
        fs_rid muri = fs_hash_uri(model[f]);
        fs_rid_vector_append(mvec, muri);
    }
    if (!adding) {
        if (verbosity) {
	    printf("removing old data\n");
	    fflush(stdout);
        }
        if (!(dryrun & FS_DRYRUN_DELETE)) {
	    if (fsp_delete_model_all(fsplink, mvec)) {
	        fs_error(LOG_ERR, "model delete failed");
	        return 1;
	    }
	    for (int i=0; i<mvec->length; i++) {
		if (mvec->data[i] == fs_c.system_config) {
		    fs_import_reread_config();
		}
	    }
        }
        fsp_new_model_all(fsplink, mvec);
    }

    fs_rid_vector_free(mvec);

    gettimeofday(&then_last, 0);
    for (int f = 0; f < files; ++f) {
	if (verbosity) {
            printf("Reading <%s>\n", uri[f]);
            if (strcmp(uri[f], model[f])) {
                printf("   into <%s>\n", model[f]);
            }
	    fflush(stdout);
        }

        fs_import(fsplink, model[f], uri[f], format, verbosity, dryrun, has_o_index, msg, &total_triples);
	if (verbosity) {
	    fflush(stdout);
        }
    }
    double sthen = fs_time();
    int ret = fs_import_commit(fsplink, verbosity, dryrun, has_o_index, msg, &total_triples);

    if (verbosity > 0) {
	printf("Updating index\n");
        fflush(stdout);
    }
    fsp_stop_import_all(fsplink);
    if (verbosity > 0) {
        printf("Index update took %f seconds\n", fs_time()-sthen);
    }

    if (!ret) {
        gettimeofday(&now, 0);
        double diff = (now.tv_sec - then.tv_sec) +
                        (now.tv_usec - then.tv_usec) * 0.000001;
        if (verbosity && total_triples > 0) {
	    printf("Imported %d triples, average %d triples/s\n", total_triples,
		     (int)((double)total_triples/diff));
            fflush(stdout);
        }
    }

    if (verbosity > 1) {
        printf("seg add_q\tadd_r\t\tcommit_q\tcommit_r\tremove\t\trebuild\t\twrite\n");
        long long *tics = fsp_profile_write(fsplink);

        for (int seg = 0; seg < segments; seg++) {
            fs_import_timing newtimes;
            fsp_get_import_times(fsplink, seg, &newtimes);

	    printf("%2d: %f\t%f\t%f\t%f\t%f\t%f\t%f\n", seg,
                   newtimes.add_s - timing[seg].add_s,
	           newtimes.add_r - timing[seg].add_r,
	           newtimes.commit_q - timing[seg].commit_q,
                   newtimes.commit_r - timing[seg].commit_r,
                   newtimes.remove - timing[seg].remove,
		   newtimes.rebuild - timing[seg].rebuild,
		   tics[seg] * 0.001);
	}
    }

    fsp_close_link(fsplink);
    raptor_free_world(rw);

    return 0;
}

/* vi:set expandtab sts=4 sw=4: */
